src/java.base/share/classes/java/lang/StableValue.java

```diff
@@ -0,0 +1,664 @@
+
+import jdk.internal.access.SharedSecrets;
+import jdk.internal.javac.PreviewFeature;
+import jdk.internal.lang.stable.StableEnumFunction;
+import jdk.internal.lang.stable.StableFunction;
+import jdk.internal.lang.stable.StableIntFunction;
+import jdk.internal.lang.stable.StableSupplier;
+import jdk.internal.lang.stable.StableValueImpl;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.RandomAccess;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.IntFunction;
+import java.util.function.Supplier;
+
+/**
+ * A stable value is a deferred holder of shallowly immutable content.
+ * <p>
+ * A {@code StableValue<T>} can be created using the factory method
+ * {@linkplain StableValue#of() {@code StableValue.of()}}. When created this way,
+ * the stable value is <em>unset</em>, which means it holds no <em>content</em>.
+ * Its content, of type {@code T}, can be <em>set</em> by calling
+ * {@linkplain #trySet(Object) trySet()}, {@linkplain #setOrThrow(Object) setOrThrow()},
+ * or {@linkplain #orElseSet(Supplier) orElseSet()}. Once set, the content
+ * can never change and can be retrieved by calling {@linkplain #orElseThrow() orElseThrow()}
+ * , {@linkplain #orElse(Object) orElse()}, or {@linkplain #orElseSet(Supplier) orElseSet()}.
+ * <p>
+ * A stable value that is <em>set</em> is treated as a constant by the JVM, enabling the
+ * same performance optimizations that are available for {@code final} fields.
+ * As such, stable values can be used to replace {@code final} fields in cases where
+ * <em>at-most-once</em> update semantics is crucial, but where the eager initialization
+ * semantics associated with {@code final} fields is too restrictive.
+ * <p>
+ * Consider the following example where a stable value field "{@code logger}" is a
+ * shallowly immutable holder of content of type {@code Logger} and that is initially
+ * created as <em>unset</em>, which means it holds no content. Later in the example, the
+ * state of the "{@code logger}" field is checked and if it is still <em>unset</em>,
+ * the content is <em>set</em>:
+ *
+ * {@snippet lang = java:
+ * public class Component {
+ *
+ *    // Creates a new unset stable value with no content
+ *    // @link substring="of" target="#of" :
+ *    private final StableValue<Logger> logger = StableValue.of();
+ *
+ *    private Logger getLogger() {
+ *        if (!logger.isSet()) {
+ *            logger.trySet(Logger.create(Component.class));
+ *        }
+ *        return logger.orElseThrow();
+ *    }
+ *
+ *    public void process() {
+ *        getLogger().info("Process started");
+ *        // ...
+ *    }
+ * }
+ *}
+ * <p>
+ * Note that the holder value can only be set at most once.
+ * In the example above, the {@code logger} field is declared {@code final} which is
```

Comment:
This is not 100% true -- a final non-static field is not treated as a constant by the VM (at least not yet)

Reply:
Now that we have removed the VM-specific handling of fields declared as `StableValue`, this is true. I will take a look at improving the wording.

Reply:
I've used `static` fields now to get out of the muddy waters of trusted final fields, records etc.

Reply:
I didn't mean to say "this is not true, so use a `static final` instead". What I meant was more "let's not dive into VM-specific details now" (see my other comment on the topic), and just focus on semantics instead. Then let's put all the constant-folding related stuff somewhere else. I liked using non-static final fields!

Reply:
Basically: drop all references to "performance" and "constant folding" everywhere. And then introduce a section dedicated to that topic which talks a bit about how the content of a SV is trusted not to change by the JVM, which enables constant-folding optimizations.

Reply:
The reason I'm suggesting this is that what's constant-folded is kind of a moving target, so we don't want to say/promise too much. For now, we can say that if the SV, or the stable function/collection is stored in a static/final field then its content will be constant-folded. But we know that there's other stuff in the pipeline (e.g. [this JEP](https://openjdk.org/jeps/8349536)) which will affect this picture, and enable for more constant-folding when accessing SV and stable functions/collections.